frontend/src/lib/utils.ts

```diff
@@ -203,3 +203,93 @@ export function labelVisibilityProtoValueToEnum(
       return LabelVisibilityOptions.Visible
   }
 }
+
+/**
+ * Looks for an IFrame with given className inside given querySet
+ */
+export function findAnIFrameWithClassName(
+  qs: NodeListOf<HTMLIFrameElement> | HTMLCollectionOf<HTMLIFrameElement>,
+  className: string
+): HTMLIFrameElement | null {
+  for (let i = 0; i < qs.length; i++) {
+    const cd = qs[i].contentDocument
+    if (cd && cd.getElementsByClassName(className).length > 0) {
+      return qs[i]
+    }
+  }
+  return null
+}
+
+/**
+ * Returns True if IFrame can be accessed otherwise returns False
+ */
+export function canAccessIFrame(iframe: HTMLIFrameElement): boolean {
+  try {
+    if (iframe.contentWindow === null) return false
+    const doc = iframe.contentDocument || iframe.contentWindow.document
+    const html = doc.body.innerHTML
+    return html !== null && html !== ""
+  } catch (err) {
+    return false
+  }
+}
+
+/**
+ * Tries to get an IFrame in which Streamlit app is embedded on Cloud deployments.
+ * It assumes iframe has title="streamlitApp", iterates over IFrames,
+ * and looks which IFrame contains div with stAppId value, otherwise returns first found iFrame or null.
+ */
+export function getIFrameEnclosingApp(
+  embeddingId: string
+): HTMLIFrameElement | null {
+  if (!isInChildFrame()) {
+    return null
+  }
+  const embeddingIdClassName = getEmbeddingIdClassName(embeddingId)
+  const qsStreamlitAppStr = 'iframe[title="streamlitApp"]'
+  let qs = window.document.querySelectorAll(
+    qsStreamlitAppStr
+  ) as NodeListOf<HTMLIFrameElement>
+  if (qs.length < 1 && window.parent) {
+    qs = window.parent.document.querySelectorAll(qsStreamlitAppStr)
+  }
+  let foundIFrame = findAnIFrameWithClassName(qs, embeddingIdClassName)
+  if (foundIFrame && !canAccessIFrame(foundIFrame)) {
+    return null
+  }
+  if (foundIFrame) {
+    return foundIFrame
+  }
+  let htmlCollection = window.document.getElementsByTagName(
+    "iframe"
+  ) as HTMLCollectionOf<HTMLIFrameElement>
+  if (htmlCollection.length < 1 && window.parent) {
+    htmlCollection = window.parent.document.getElementsByTagName("iframe")
+  }
+  if (htmlCollection.length > 0) {
+    foundIFrame = htmlCollection[0] // eslint-disable-line prefer-destructuring
```

Comment:
I think we'll want to also look through all of the iframes for one with the class we're expecting in this case too (although if both cases are generalizable into a helper single function, this will happen for for free)

Reply:
I've done it on purpose, please check out my larger comment above

Reply:
🤔 at first when I was implementing this I wanted to avoid this and just return first IFrame if strict looking fails, however I think it's actually better to look and iterate whenever possible, now I look for an `embeddedId` whenever I have any kind of queryset, so I've updated it, thank you 🙇